### PR TITLE
[eclipse/xtext#1569] Refactor Xtend to Java

### PR DESCRIPTION
**Pull request by kthoms**
The usage of Xtend in the Xtext code base should be reduced where the
usage of Xtend is not super beneficial. Replaced Xtend classes by the
compiled code and refactored the code to more readable Java.

Signed-off-by: Karsten Thoms <karsten.thoms@itemis.de>

### CODE REVIEW

File: org.eclipse.xtext/src/org/eclipse/xtext/parser/antlr/AbstractAntlrParserBasedTokenSourceProvider.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+public class AbstractAntlrParserBasedTokenSourceProvider extends AbstractTokenSourceProvider {
+	@Inject
+	private IParser parser;
+
+	@Override
+	public TokenSource createTokenSource(final CharStream stream) {
+		if (this.parser instanceof AbstractAntlrParser) {
+			return ((AbstractAntlrParser) this.parser).createLexer(stream);
+		}
+		StringBuilder builder = new StringBuilder()
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/parser/antlr/AbstractAntlrParserBasedTokenSourceProvider.java]: Unnecessary usage of StringBuilder. String concatenation is more readable.

File: org.eclipse.xtext/src/org/eclipse/xtext/parser/antlr/AbstractTokenSourceProvider.java [hunk 2]
```diff
@@ -9,38 +9,36 @@
...
-  }
-  
-  @Override
-  public TokenSource createTokenSource(final Reader reader) {
-    return this.createTokenSource(this.getCharStream(reader));
-  }
+	protected CharStream getCharStream(final Reader reader) {
+		try {
+			return new ANTLRReaderStream(reader);
+		} catch (Throwable e) {
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/parser/antlr/AbstractTokenSourceProvider.java]: Please do not use sneaky throw

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/ForwardingEObjectDescription.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+ * Suitable to override and specialize behavior.
+ * 
+ * @author Sebastian Zarnekow - Initial contribution and API
+ * @since 2.9
+ */
+public abstract class ForwardingEObjectDescription extends ForwardingObject implements IEObjectDescription {
+	private final IEObjectDescription delegate;
+
+	public ForwardingEObjectDescription(final IEObjectDescription delegate) {
+		super();
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/ForwardingEObjectDescription.java]: Unnecessary `super();`

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/ForwardingEObjectDescription.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+	}
+
+	@Override
+	protected IEObjectDescription delegate() {
+		return this.delegate;
+	}
+
+	@Override
+	public String toString() {
+		return (("ForwardingEObjectDescription[" + this.delegate) + "]");
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/ForwardingEObjectDescription.java]: Unnecessary parentheses

File: org.eclipse.xtext/src/org/eclipse/xtext/service/AllRulesCache.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+		}
+	}
+
+	private final Map<String, AbstractRule> allRules;
+
+	public AllRulesCache(final Grammar grammar) {
+		allRules = Maps.newHashMap();
+		List<AbstractRule> allRules = GrammarUtil.allRules(grammar);
+		for (final AbstractRule rule : allRules) {
+			if ((rule instanceof TerminalRule)) {
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/service/AllRulesCache.java]: Unnecessary parentheses. Please use code cleanup actions on the ported code.

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/IMimeTypeProvider.java [hunk 6]
```diff
@@ -7,23 +7,23 @@
...
-    @Override
-    public List<String> getMimeTypes() {
-      return Collections.<String>unmodifiableList(CollectionLiterals.<String>newArrayList());
-    }
-  }
-  
-  public abstract List<String> getMimeTypes();
+	List<String> getMimeTypes();
+
+	class DefaultImpl implements IMimeTypeProvider {
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/resource/IMimeTypeProvider.java]: no more need for public?

File: org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledError.java [hunk 7]
```diff
@@ -10,14 +10,13 @@
 /**
  * @since 2.8
  */
-@SuppressWarnings("all")
+@SuppressWarnings("serial")
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledError.java]: Is this about the missing serialization id? Might be better to generate one then

File: org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledError.java [hunk 8]
```diff
@@ -10,14 +10,13 @@
...
 public class OperationCanceledError extends Error {
-  public OperationCanceledError(final RuntimeException cause) {
-    super(cause);
-  }
-  
-  public RuntimeException getWrapped() {
-    Throwable _cause = this.getCause();
-    return ((RuntimeException) _cause);
-  }
+	public OperationCanceledError(final RuntimeException cause) {
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledError.java]: Seems unneeded to me

File: org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledManager.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+	public boolean isOperationCanceledException(final Throwable t) {
+		return (getPlatformOperationCanceledException(t) != null);
+	}
+
+	/**
+	 * Rethrows OperationCanceledErrors and wraps platform specific OperationCanceledExceptions. Does nothing for any
+	 * other type of Throwable.
+	 */
+	public void propagateAsErrorIfCancelException(final Throwable t) {
+		if ((t instanceof OperationCanceledError)) {
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledManager.java]: Cleanup actions <3

File: org.eclipse.xtext/src/org/eclipse/xtext/service/AllRulesCache.java (hunk 5 above)
Comment (kthoms) [reply org.eclipse.xtext/src/org/eclipse/xtext/service/AllRulesCache.java]: cleanup actions do not cover that :( I've already performed them

File: org.eclipse.xtext/src/org/eclipse/xtext/parser/antlr/AbstractAntlrParserBasedTokenSourceProvider.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+ * @author kosyakov - Initial contribution and API
+ * @since 2.8
+ */
+@Singleton
+public class AbstractAntlrParserBasedTokenSourceProvider extends AbstractTokenSourceProvider {
+	@Inject
+	private IParser parser;
+
+	@Override
+	public TokenSource createTokenSource(final CharStream stream) {
```
[the remark this answers is not in the document]
Comment (kthoms) [reply org.eclipse.xtext/src/org/eclipse/xtext/parser/antlr/AbstractAntlrParserBasedTokenSourceProvider.java]: No they can't. You can only add final, but not remove them with Cleanup.

File: org.eclipse.xtext/src/org/eclipse/xtext/resource/IMimeTypeProvider.java (hunk 6 above)
Comment (kthoms) [reply org.eclipse.xtext/src/org/eclipse/xtext/resource/IMimeTypeProvider.java]: public is implicit in interfaces, and static for classes too

File: org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledError.java (hunk 7 above)
Comment (kthoms) [reply org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledError.java]: we don't serialize anything, but why not. I've added it.

File: org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledManager.java (hunk 9 above)
Comment (kthoms) [reply org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledManager.java]: They don't remove this :(

File: org.eclipse.xtext/src/org/eclipse/xtext/service/AllRulesCache.java [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+	}
+
+	private final Map<String, AbstractRule> allRules;
+
+	public AllRulesCache(Grammar grammar) {
+		allRules = Maps.newHashMap();
+		List<AbstractRule> allRules = GrammarUtil.allRules(grammar);
+		for (AbstractRule rule : allRules) {
+			if (rule instanceof TerminalRule) {
+				this.allRules.put(((TerminalRule) rule).getName().toUpperCase(), rule);
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/service/AllRulesCache.java]: Renaming `this.allRules` to `allRulesCache` might be better

File: org.eclipse.xtext/src/org/eclipse/xtext/service/AllRulesCache.java [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+			}
+		}
+		return null;
+	}
+
+	public static AllRulesCache removeFromEmfObject(Notifier emfObject) {
+		List<Adapter> adapters = emfObject.eAdapters();
+		for (int i = 0, max = adapters.size(); i < max; i++) {
+			Adapter adapter = adapters.get(i);
+			if (adapter instanceof AllRulesCache.AllRulesCacheAdapter) {
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/service/AllRulesCache.java]: No need for `AllRulesCache.`

File: org.eclipse.xtext/src/org/eclipse/xtext/service/AllRulesCache.java [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+		}
+		return null;
+	}
+
+	public void attachToEmfObject(Notifier emfObject) {
+		AllRulesCache result = findInEmfObject(emfObject);
+		if (result != null) {
+			throw new IllegalStateException("The given EMF object already contains an adapter for AllRulesCache");
+		}
+		AllRulesCache.AllRulesCacheAdapter adapter = new AllRulesCache.AllRulesCacheAdapter(this);
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/service/AllRulesCache.java]: No need for `AllRulesCache.`

File: org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledError.java [hunk 14]
```diff
@@ -10,14 +10,14 @@
...
-    return ((RuntimeException) _cause);
-  }
+	private static final long serialVersionUID = 3488104742236941451L;
+
+	public OperationCanceledError(RuntimeException cause) {
+		super(cause);
+	}
+
+	public RuntimeException getWrapped() {
+		return ((RuntimeException) getCause());
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledError.java]: Unneeded parenthesis

File: org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledManager.java [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+ * A facade for managing and working with cancellation exceptions of different platforms.
+ *
+ * @author Sven Efftinge - Initial contribution and API
+ *
+ * @since 2.8
+ */
+public class OperationCanceledManager {
+	protected RuntimeException getPlatformOperationCanceledException(Throwable t) {
+		if (t instanceof OperationCanceledException) {
+			return ((RuntimeException) t);
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledManager.java]: Unneeded parenthesis

File: org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledManager.java [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+		return getPlatformOperationCanceledException(t) != null;
+	}
+
+	/**
+	 * Rethrows OperationCanceledErrors and wraps platform specific OperationCanceledExceptions. Does nothing for any
+	 * other type of Throwable.
+	 */
+	public void propagateAsErrorIfCancelException(Throwable t) {
+		if (t instanceof OperationCanceledError) {
+			throw ((OperationCanceledError) t);
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledManager.java]: Unneeded parenthesis

File: org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledManager.java [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+	public void throwOperationCanceledException() {
+		throw asWrappingOperationCanceledException(getPlatformSpecificOperationCanceledException());
+	}
+
+	protected RuntimeException getPlatformSpecificOperationCanceledException() {
+		return new OperationCanceledException();
+	}
+
+	public void checkCanceled(final CancelIndicator indicator) {
+		if ((indicator != null) && indicator.isCanceled()) {
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledManager.java]: Unneeded parenthesis

**Comment by szarnekow** [comment]
There should be a cleanup action that allows to remove redundant parentheses and redundant modifiers.

**Comment by szarnekow** [comment]
Screenshot 2019-10-30 at 16 45 43
Screenshot 2019-10-30 at 16 45 53
Screenshot 2019-10-30 at 16 45 21
Screenshot 2019-10-30 at 16 45 31

**Comment by kthoms** [comment]
The "remove redundant modifiers" action does not remove final modifiers from parameters. How could this be decided that it might not be wanted? It removes e.g. 'public' on interface fields/methods.  BTW: I have implemented this cleanup action :)
But I did not use the "Use parentheses in expressions / Only if necessary" yet.

I think we should configure those common cleanup actions on the projects with separate issue.

**Comment by szarnekow** [comment]
> I think we should configure those common cleanup actions on the projects with separate issue.

No I don't think so. When we migrate the stuff to Java, the result should look nice and not like a second class citizen.

**Comment by kthoms** [comment]
Right. What I meant is: When we already agreed which cleanup actions make sense in general, then we could store these settings in the projects. For the migration it is clear that the resulting Java code should look nice as hand crafted, and applying the cleanup actions is just task to achieve this.

File: org.eclipse.xtext/src/org/eclipse/xtext/service/AllRulesCache.java [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+	}
+
+	private final Map<String, AbstractRule> allRulesCache;
+
+	public AllRulesCache(Grammar grammar) {
+		allRulesCache = Maps.newHashMap();
+		List<AbstractRule> allRules = GrammarUtil.allRules(grammar);
+		for (AbstractRule rule : allRules) {
+			if (rule instanceof TerminalRule) {
+				allRulesCache.put(((TerminalRule) rule).getName().toUpperCase(), rule);
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/service/AllRulesCache.java]: Unnecessary cast

File: org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledManager.java [hunk 19]
```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright (c) 2014, 2016 itemis AG (http://www.itemis.eu) and others.
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledManager.java]: Shouldn't the dates on this be updated as well?

File: org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledManager.java [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+		return getPlatformOperationCanceledException(t) != null;
+	}
+
+	/**
+	 * Rethrows OperationCanceledErrors and wraps platform specific OperationCanceledExceptions. Does nothing for any
+	 * other type of Throwable.
+	 */
+	public void propagateAsErrorIfCancelException(Throwable t) {
+		if (t instanceof OperationCanceledError) {
+			throw (OperationCanceledError) t;
```
Comment (tivervac) [inline org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledManager.java]: Unnecessary cast

File: org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledManager.java (hunk 19 above)
Comment (kthoms) [reply org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledManager.java]: Correct. I'll update this for this change set and consider it for more changes to come.

File: org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledManager.java (hunk 20 above)
Comment (kthoms) [reply org.eclipse.xtext/src/org/eclipse/xtext/service/OperationCanceledManager.java]: It is necessary here. t is a Throwable, which could be a checked exception also.

**Comment by kthoms** [comment]
@szarnekow I think all issues should be adressed now. Could you have a look again, please?

**Pull request merged**